test/common_node_test.cpp

```diff
@@ -0,0 +1,263 @@
+
+    rclcpp::Subscription<interfaces::msg::JobFinished>::SharedPtr job_finished_sub = test_node->create_subscription<interfaces::msg::JobFinished>(
+        "job_finished",
+        10,
+        [common_node, test_node, &executor](interfaces::msg::JobFinished::ConstSharedPtr msg)
+        {
+            RCLCPP_DEBUG(test_node->get_logger(), "Got job_finished message");
+            ASSERT_EQ(msg->sender_id, "/common_node");
+            ASSERT_EQ(msg->error_code, EXIT_FAILURE);
+
+            nlohmann::json payload_check;
+            payload_check["error_msg"] = "This is my error message";
+            ASSERT_EQ(payload_check, nlohmann::json::parse(msg->payload));
+
+            ASSERT_FALSE(common_node->get_active());
+            executor.cancel();
+        });
+
+    rclcpp::TimerBase::SharedPtr job_finished_timer = common_node->create_wall_timer(
+        std::chrono::milliseconds(100), [common_node]()
+        {
+            RCLCPP_DEBUG(common_node->get_logger(), "Sending job_finished message");
+            common_node->job_finished("This is my error message"); });
+
+    executor.add_node(common_node);
+    executor.add_node(test_node);
+
+    executor.spin();
+}
+
+TEST(common_package, job_finished_custom_payload)
+{
+    rclcpp::executors::SingleThreadedExecutor executor;
+
+    std::shared_ptr<JobFinishedCommonNode> common_node = std::make_shared<JobFinishedCommonNode>("common_node");
+    rclcpp::Node::SharedPtr test_node = std::make_shared<rclcpp::Node>("test");
+
+    rclcpp::Subscription<interfaces::msg::JobFinished>::SharedPtr job_finished_sub = test_node->create_subscription<interfaces::msg::JobFinished>(
+        "job_finished",
+        10,
+        [common_node, test_node, &executor](interfaces::msg::JobFinished::ConstSharedPtr msg)
+        {
+            RCLCPP_DEBUG(test_node->get_logger(), "Got job_finished message");
+            ASSERT_EQ(msg->sender_id, "/common_node");
+            ASSERT_EQ(msg->error_code, EXIT_FAILURE);
+
+            nlohmann::json payload_check;
+            payload_check["coord"] = "These are my coordinates";
+            payload_check["height_cm"] = 500;
+            ASSERT_EQ(payload_check, nlohmann::json::parse(msg->payload));
+
+            ASSERT_FALSE(common_node->get_active());
+            executor.cancel();
+        });
+
+    rclcpp::TimerBase::SharedPtr job_finished_timer = common_node->create_wall_timer(
+        std::chrono::milliseconds(100), [common_node]()
+        {
+            RCLCPP_DEBUG(common_node->get_logger(), "Sending job_finished message");
+            nlohmann::json payload;
+            payload["coord"] = "These are my coordinates";
+            payload["height_cm"] = 500;
+            common_node->job_finished(EXIT_FAILURE, payload); });
+
+    executor.add_node(common_node);
+    executor.add_node(test_node);
+
+    executor.spin();
+}
+
+int main(int argc, char **argv)
+{
+    rclcpp::init(0, nullptr);
```

Comment:
Es würde vielleicht Sinn machen, die Tests in einer unique domain laufen zu lassen.
Damit es nicht zu Problemen kommt, wenn bereits ein ROS System in der default domain unterwegs ist.
(Worstcase wir testen während eine Drohne fliegt und die teilen sich das Netz).
```suggestion
    rclcpp::InitOptions init_options;
    init_options.set_domain_id(2);
    rclcpp::init(0, nullptr, init_options);
```
Eventuell ist noch ein `#include "rclcpp/init_options.hpp"` nötig.

* https://docs.ros.org/en/humble/p/rclcpp/generated/function_namespacerclcpp_1a026b2ac505c383735117de5d1679ed80.html
* https://docs.ros.org/en/humble/p/rclcpp/generated/classrclcpp_1_1InitOptions.html#_CPPv4N6rclcpp11InitOptionsE
* https://docs.ros.org/en/humble/Concepts/Intermediate/About-Domain-ID.html

Reply:
Kann man bei colcon so Builden, dass die Tests nicht mitgebuildet werden?
Dann könnte man dieses Szenario zumindest etwas unwahrscheinlicher machen.

Reply:
Andere Domain ist btw. eingebaut.

Reply:
> Kann man bei colcon so Builden, dass die Tests nicht mitgebuildet werden? Dann könnte man dieses Szenario zumindest etwas unwahrscheinlicher machen.

@J0las Weißt du das?

Reply:
Mit diesem argument sollte er keine Tests mehr builden: `--cmake-args -DBUILD_TESTING=OFF`
Ich würde das halt wegen der theoretischen Möglichkeit machen, das man am Flugtag im Feld changes macht, diese dann testet, während eine Drohne fliegt und man im selben Netz ist.
Da hat man vllt vergessen, das die sich dann die ROS Domain teilen und dann kommt es zu Problemen, die einen verwirren.
So sollte das auszuschließen sein. Das DDS unter ROS nutzt dann andere Ports.

Reply:
Da war ich gerade am schreiben😅

> > Kann man bei colcon so Builden, dass die Tests nicht mitgebuildet werden? Dann könnte man dieses Szenario zumindest etwas unwahrscheinlicher machen.
> 
> @J0las Weißt du das?

Reply:
Ich hoffe nicht, dass dieser Fall eintritt 😅.
Ich werden das Argument mal notieren, damit man die Version auf der Drohne am Besten ohne Test buildet. Dann können die zumindest schonmal nicht dort ausgeführt werden.